concert/app/src/functions/authorizer/index.py

```diff
@@ -1,36 +1,96 @@
 import base64
+import hashlib
 import json
 import os
+import time
+from jose import jwk, jwt
+from jose.utils import base64url_decode
+from urllib.request import urlopen
 
 API_ARN = os.environ.get("API_ARN")
+API_METHOD_CHOICE_MADE = os.environ.get("API_METHOD_CHOICE_MADE")
+API_METHOD_EVENT_STAGE_CHANGED = os.environ.get("API_METHOD_EVENT_STAGE_CHANGED")
+COGNITO_APP_CLIENT_ID = os.environ.get("COGNITO_APP_CLIENT_ID")
+COGNITO_GROUP_ADMIN = os.environ.get("COGNITO_GROUP_ADMIN")
+COGNITO_GROUP_VISUALIZATION = os.environ.get("COGNITO_GROUP_VISUALIZATION")
 COGNITO_USER_POOL_ID = os.environ.get("COGNITO_USER_POOL_ID")
-COGNITO_ADMIN_GROUP_NAME = os.environ.get("COGNITO_ADMIN_GROUP_NAME")
+COGNITO_REGION = COGNITO_USER_POOL_ID.split("_")[0]
 
-# TODO:
-# Verify JWT
-# Check that user is in the concert group
-# Write method-specific policy based on user groups (admin, vis, etc.)
+response = urlopen(
+    "https://cognito-idp.{}.amazonaws.com/{}/.well-known/jwks.json".format(
+        COGNITO_REGION, COGNITO_USER_POOL_ID
+    )
+)
+JWKS_KEYS = json.loads(response.read())["keys"]
 
+CHOICE_BUCKETS = [
+    "CHOICE_COLOR",
+    "CHOICE_COLOR",
+    "CHOICE_EMOTION_HAPPINESS",
+    "CHOICE_EMOTION_AGITATION",
+]
+NUM_BUCKETS = len(CHOICE_BUCKETS)
 
-def handler(event, context):
-    token = event["headers"]["Authorization"]
-    payload = {"sub": token}
 
-    try:
-        payload = json.loads(base64.decode(token.split(".")[1]))
-    except Exception as e:
-        print(e)
+def get_verified_token(token):
+    headers = jwt.get_unverified_headers(token)
+    kid = headers["kid"]
+
+    key_index = -1
+    for i in range(len(JWKS_KEYS)):
+        if kid == JWKS_KEYS[i]["kid"]:
+            key_index = i
+            break
+    if key_index == -1:
+        return None
+
+    return jwt.decode(token, JWKS_KEYS[key_index], audience=COGNITO_APP_CLIENT_ID)
+
+
+def get_user_context(user_id, groups):
```

Comment:
It's not obvious to me why we're doing the modular arithmetic below in the way that we're doing it. Would you mind explaining?

Reply:
We need two random variables: choiceType and choiceInverted. 50% of people should be color, 25% happiness, 25% agitation. That's the first modulo. We also want 50% split for inverted/not. That's the second modulo. So there are eight groups: four are color, two are happiness (inverted/not), two are agitation (inverted/not).

Reply:
Roger. Thanks for the explanation.